src/appleseed/foundation/math/aabb.h

```diff
@@ -465,75 +425,118 @@ inline size_t AABB<T, N>::rank() const
 }
 
 template <typename T, size_t N>
-inline Vector<T, N> AABB<T, N>::center() const
+inline size_t AABBBase<T, N>::get_corner_count() const
+{
+    return 1 << N;
+}
+
+template <typename T, size_t N>
+Vector<T, N> AABBBase<T, N>::compute_corner(const size_t i) const
 {
     assert(is_valid());
 
-    return ValueType(0.5) * (min + max);
+    VectorType p;
+
+    for (size_t d = 0; d < N; ++d)
+        p[d] = i & (size_t(1) << d) ? max[d] : min[d];
+
+    return p;
 }
 
 template <typename T, size_t N>
-inline T AABB<T, N>::center(const size_t dim) const
+inline bool AABBBase<T, N>::contains(const VectorType& v) const
 {
     assert(is_valid());
 
-    return ValueType(0.5) * (min[dim] + max[dim]);
+    for (size_t i = 0; i < N; ++i)
+    {
+        if (v[i] < min[i] || v[i] > max[i])
+            return false;
+    }
+
+    return true;
 }
 
 template <typename T, size_t N>
-inline Vector<T, N> AABB<T, N>::extent() const
+inline bool operator!=(const AABBBase<T, N>& lhs, const AABBBase<T, N>& rhs)
 {
-    assert(is_valid());
+    return lhs.min != rhs.min || lhs.max != rhs.max;
+}
 
-    return max - min;
+template <typename T, size_t N>
+inline bool operator==(const AABBBase<T, N>& lhs, const AABBBase<T, N>& rhs)
+{
+    return !(lhs != rhs);
 }
 
 template <typename T, size_t N>
-inline T AABB<T, N>::extent(const size_t dim) const
+inline bool feq(const AABBBase<T, N>& lhs, const AABBBase<T, N>& rhs)
```

Comment:
We probably don't want to define `feq()` (approximate equality test) for integral AABBs.